src/main/java/com/google/cloud/tools/appengine/cloudsdk/GcloudRunner.java

```diff
@@ -0,0 +1,150 @@
+  private final ProcessBuilderFactory processBuilderFactory;
+  private final ProcessHandler processHandler;
+
+  GcloudRunner(
+      CloudSdk sdk,
+      String metricsEnvironment,
+      String metricsEnvironmentVersion,
+      File credentialFile,
+      String outputFormat,
+      String showStructuredLogs,
+      ProcessBuilderFactory processBuilderFactory,
+      ProcessHandler processHandler) {
+    this.sdk = sdk;
+    this.metricsEnvironment = metricsEnvironment;
+    this.metricsEnvironmentVersion = metricsEnvironmentVersion;
+    this.credentialFile = credentialFile;
+    this.outputFormat = outputFormat;
+    this.showStructuredLogs = showStructuredLogs;
+    this.processBuilderFactory = processBuilderFactory;
+    this.processHandler = processHandler;
+  }
+
+  void run(List<String> arguments, File workingDirectory)
+      throws ProcessHandlerException, CloudSdkNotFoundException, CloudSdkOutOfDateException,
+          CloudSdkVersionFileException, InvalidJavaSdkException, IOException {
+
+    sdk.validateCloudSdk();
+
+    List<String> command = new ArrayList<>();
+    command.add(sdk.getGCloudPath().toString());
+
+    command.addAll(arguments);
+    command.addAll(GcloudArgs.get("format", outputFormat));
+
+    if (credentialFile != null) {
+      command.addAll(GcloudArgs.get("credential-file-override", credentialFile));
+    }
+
+    logger.info("submitting command: " + Joiner.on(" ").join(command));
+
+    ProcessBuilder processBuilder = processBuilderFactory.newProcessBuilder();
+    processBuilder.command(command);
+    if (workingDirectory != null) {
+      processBuilder.directory(workingDirectory);
+    }
+    processBuilder.environment().putAll(getGcloudCommandEnvironment());
+    Process process = processBuilder.start();
+    processHandler.handleProcess(process);
+  }
+
+  @VisibleForTesting
+  Map<String, String> getGcloudCommandEnvironment() {
+    Map<String, String> environment = Maps.newHashMap();
+    if (credentialFile != null) {
+      environment.put("CLOUDSDK_APP_USE_GSUTIL", "0");
+    }
+    if (metricsEnvironment != null) {
+      environment.put("CLOUDSDK_METRICS_ENVIRONMENT", metricsEnvironment);
+    }
+    if (metricsEnvironmentVersion != null) {
+      environment.put("CLOUDSDK_METRICS_ENVIRONMENT_VERSION", metricsEnvironmentVersion);
+    }
+    if (showStructuredLogs != null) {
+      environment.put("CLOUDSDK_CORE_SHOW_STRUCTURED_LOGS", showStructuredLogs);
+    }
+    // This is to ensure IDE credentials get correctly passed to the gcloud commands, in Windows.
+    // It's a temporary workaround until a fix is released.
+    // https://github.com/GoogleCloudPlatform/google-cloud-intellij/issues/985
+    if (System.getProperty("os.name").contains("Windows")) {
+      environment.put("CLOUDSDK_APP_NUM_FILE_UPLOAD_PROCESSES", "1");
+    }
+
+    environment.put("CLOUDSDK_CORE_DISABLE_PROMPTS", "1");
+
+    return environment;
+  }
+
+  static class Factory {
```

Comment:
It seems like this `Factory` is not necessary as well since it just wraps another factory.

Reply:
Yeah, similar to the AppCfg thing,